mp2-v1/src/values_extraction/leaf_receipt.rs

```diff
@@ -0,0 +1,542 @@
+//! Module handling the leaf node inside a Receipt Trie
+
+use crate::MAX_RECEIPT_LEAF_NODE_LEN;
+
+use super::public_inputs::{PublicInputs, PublicInputsArgs};
+
+use mp2_common::{
+    array::{Array, Vector, VectorWire},
+    eth::{EventLogInfo, LogDataInfo, ReceiptProofInfo},
+    group_hashing::CircuitBuilderGroupHashing,
+    keccak::{InputData, KeccakCircuit, KeccakWires},
+    mpt_sequential::{MPTKeyWire, MPTReceiptLeafNode, PAD_LEN},
+    public_inputs::PublicInputCommon,
+    rlp::MAX_KEY_NIBBLE_LEN,
+    types::{CBuilder, GFp},
+    utils::{less_than, less_than_or_equal_to, Endianness, PackerTarget},
+    D, F,
+};
+use plonky2::{
+    field::types::Field,
+    iop::{
+        target::Target,
+        witness::{PartialWitness, WitnessWrite},
+    },
+    plonk::circuit_builder::CircuitBuilder,
+};
+
+use plonky2_ecgfp5::gadgets::curve::{CircuitBuilderEcGFp5, CurveTarget};
+
+use recursion_framework::circuit_builder::CircuitLogicWires;
+use rlp::Encodable;
+use serde::{Deserialize, Serialize};
+use std::{array::from_fn, iter};
+/// Maximum number of logs per transaction we can process
+const MAX_LOGS_PER_TX: usize = 2;
+
+#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
+pub struct ReceiptLeafWires<const NODE_LEN: usize>
+where
+    [(); PAD_LEN(NODE_LEN)]:,
+{
+    /// The event we are monitoring for
+    pub event: EventWires,
+    /// The node bytes
+    pub node: VectorWire<Target, { PAD_LEN(NODE_LEN) }>,
+    /// the hash of the node bytes
+    pub root: KeccakWires<{ PAD_LEN(NODE_LEN) }>,
+    /// The index of this receipt in the block
+    pub index: Target,
+    /// The offset of the status of the transaction in the RLP encoded receipt node.
+    pub status_offset: Target,
+    /// The offsets of the relevant logs inside the node
+    pub relevant_logs_offset: VectorWire<Target, MAX_LOGS_PER_TX>,
+    /// The key in the MPT Trie
+    pub mpt_key: MPTKeyWire,
+}
+
+/// Contains all the information for an [`Event`] in rlp form
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
+pub struct EventWires {
+    /// Size in bytes of the whole event
+    size: Target,
+    /// Packed contract address to check
+    address: Array<Target, 20>,
+    /// Byte offset for the address from the beginning of a Log
+    add_rel_offset: Target,
+    /// Packed event signature,
+    event_signature: Array<Target, 32>,
+    /// Byte offset from the start of the log to event signature
+    sig_rel_offset: Target,
+    /// The topics for this Log
+    topics: [LogColumn; 3],
+    /// The extra data stored by this Log
+    data: [LogColumn; 2],
```

Comment:
For `data` yes we probably should, but `topics` is hard limited by solidity at 4 (with the first being the event signature hash) so I think it should be fine to leave this as three rather than add another constant.